src/peft/tuners/xlora/model.py

```diff
@@ -0,0 +1,368 @@
+
+                del kwargs_real["_xlora_classifier_inhibitor_flag"]
+
+                self.internal_xlora_scalings = torch.full(  # type: ignore
+                    (payload.batch_size, payload.seq_len, xlora_classifier.n_layers, xlora_classifier.n_classes),
+                    payload.override_scaling_pass_value,
+                )
+
+                return
+
+            xlora_scalings = xlora_classifier.forward(
+                *args_real,
+                **kwargs_real,
+            )
+            # Set the scalings
+            self.internal_xlora_scalings = xlora_scalings
+
+        model.register_forward_pre_hook(hook, with_kwargs=True, prepend=True)
+
+        self._freeze_all_adapters()
+
+        total_swapped = convert_layers_to_xlora(
+            model_peft,
+            self,
+            peft_config,
+        )
+
+        n_classes = len(peft_config.adapters)
+        xlora_classifier = XLoraClassifier(model_peft, peft_config, n_classes, total_swapped)
+
+        # Setup the model internal state
+        self.__dict__["internal_xlora_classifier"] = xlora_classifier
+        self.internal_xlora_scalings = None  # type: ignore
+
+    def _freeze_all_adapters(self):
+        self.eval()
+        if not self.xlora_config.use_trainable_adapters:
+            for name, param in self.named_parameters():
+                if "lora_" in name:
+                    param.requires_grad = False
+
+    def generate(self, *args, **kwargs):
+        # Rely on LoraModel.__getattr__
+        res = self.model.generate(*args, **kwargs)  # type: ignore
+        # TODO(EricLBuehler): Evaluate effectiveness and performance degradation
+        self._freeze_all_adapters()
+        return res
+
+    def _mark_only_adapters_as_trainable(self, model: nn.Module) -> None:
+        active_adapters = []
+        copy = self.active_adapters.copy()
+        for name in self.active_adapters:
+            if not isinstance(self.peft_config[name], XLoraConfig):
+                active_adapters.append(name)
+        self.active_adapter = active_adapters
+        super()._mark_only_adapters_as_trainable(model)
+
+        self.active_adapter = copy
+
+    def _save_pretrained_hook(
+        self,
+        save_directory: str,
+        safe_serialization: bool = True,
+        is_main_process: bool = True,
+        **kwargs: Any,
+    ) -> None:
+        classifier: XLoraClassifier = self.internal_xlora_classifier
+
+        conf = self.xlora_config.__dict__.copy()
+        del conf["device"]
+
+        conf["adapters"] = list(conf["adapters"].keys())
+        with open(os.path.join(save_directory, "xlora_config.json"), "w") as f:
+            json.dump(conf, f)
+
+        if safe_serialization:
+            # https://github.com/huggingface/peft/blob/main/src/peft/peft_model.py#L223
+            if is_main_process and safe_serialization:
+                save_model(classifier, os.path.join(save_directory, "xlora_classifier.safetensors"))
```

Comment:
This is necessary because the XLoraClassifier stores the base model, so would cause a recursion error if it was properly registered with Pytorch. To get around this, we manually assign the attribute [here](https://github.com/huggingface/peft/pull/1491/files#diff-3c6d16997aeea2a67700bdda341de1ab5161294c2fd183d583961c0c028de1d7R211).